sefaria/model/text.py

```diff
@@ -1686,6 +1739,76 @@ def __call__(cls, *args, **kwargs):
             return super(TextFamilyDelegator, cls).__call__(*args, **kwargs)
 
 
+class TextRange:
+
+    def __init__(self, oref, lang, vtitle, merge_versions=False):
+        if isinstance(oref.index_node, JaggedArrayNode) or isinstance(oref.index_node, DictionaryEntryNode): #text cannot be SchemaNode
+            self.oref = oref
+        elif oref.has_default_child(): #use default child:
+            self.oref = oref.default_child_ref()
+        else:
+            raise InputError("Can not get TextRange at this level, please provide a more precise reference")
+        self.lang = lang
+        self.vtitle = vtitle
+        self.merge_versions = merge_versions
+        self._versions = []
+        self._text = None
+        self.sources = None
+
+    @property
+    def versions(self):
+        if self._versions == []:
+            condition_query = self.oref.condition_query(self.lang) if self.merge_versions else \
+                {'title': self.oref.index.title, 'actualLanguage': self.lang, 'versionTitle': self.vtitle}
+            self._versions = VersionSet(condition_query, proj=self.oref.part_projection())
+        return self._versions
+
+    @versions.setter
+    def versions(self, versions):
+        self._validate_versions(versions)
+        self._versions = versions
+
+    def _validate_versions(self, versions):
+        if not self.merge_versions and len(versions) > 1:
+            raise InputError("Got many versions instead of one")
+        for version in versions:
+            condition = version.title == self.oref.index.title and version.actualLanguage == self.lang
+            if not self.merge_versions:
+                condition = condition and version.versionTitle == self.vtitle
+            if not condition:
+                raise InputError(f"Given version, {version}, is not matching to title, language or versionTitle")
+
+    def _trim_text(self, text):
+        """
+        part_projection trims only the upper level of the jagged array. this function trims its lower levels and get rid of 1 element arrays wrappings
+        """
+        #TODO can we get the specific text directly from mongo?
+        text = copy.deepcopy(text)
+        for s, section in enumerate(self.oref.toSections[1:], 1): #start cut from end, for cutting from the start will change the indexes
+            subtext = reduce(lambda x, _: x[-1], range(s), text)
+            del subtext[section:]
+        for s, section in enumerate(self.oref.sections[1:], 1):
+            subtext = reduce(lambda x, _: x[0], range(s), text)
+            del subtext[:section-1]
+        matching_sections = itertools.takewhile(lambda pair: pair[0] == pair[1], zip(self.oref.sections, self.oref.toSections))
+        redundant_depth = len(list(matching_sections))
+        return reduce(lambda x, _: x[0], range(redundant_depth), text)
+
+    @property
```

Comment:
This is great, will allow us the flexibility to return a flattened version of the text as well in the future